metrics/src/metric_types.rs

```diff
@@ -0,0 +1,110 @@
+use std::sync::atomic::{AtomicU64, Ordering};
+
+/// Mimics a [`metrics-core`] monotonically increasing [`Counter`] type
+pub struct Counter(AtomicU64);
+
+impl Counter {
+    pub(crate) const fn new() -> Self {
+        Self(AtomicU64::new(0))
+    }
+
+    /// Increases the value of the [`Counter`] by a discrete amount
+    #[inline]
+    pub(crate) fn increment(&self, val: u64) {
+        self.0.fetch_add(val, Ordering::SeqCst);
```

Comment:
```suggestion
        self.0.fetch_add(val, Ordering::Relaxed);
```
This counter can only ever increase, so `Relaxed` ordering should be sufficient here, no?

Reply:
You can still get read-after-write data races using `Relaxed` ordering. However I do think we can reduce the guarantees down to `Release`, as snapshot reads are always relaxed. The same for the Gauges.

Reply:
It's just a nit though; we can be conservative for now and adjust later if need be 👍.

Reply:
That's fair. Given the uncertainty towards the correctness of the metrics, being conservative is the better choice. I had already relaxed them down to `Release`, so I'm happy to leave them as-is :-)